cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/BodyConsumerAdapter.java

```diff
@@ -160,51 +160,40 @@ public void run(DatasetContext context) throws Exception {
       responder.setTransactionFailureResponse(t);
     } finally {
       try {
-        serviceContext.dismissTransactionContext();
         responder.execute(false);
       } finally {
-        contextReleaser.cancel();
-      }
-    }
-  }
-
-  private Transactional createTransactional(final TransactionContext txContext) {
-    return new Transactional() {
-      @Override
-      public void execute(final TxRunnable runnable) throws TransactionFailureException {
-        // This method is only called from user code, hence the context classloader is the
-        // program context classloader (or whatever the user set to).
-        // We need to switch the classloader back to CDAP system classloader before calling txExecute
-        // since it starting of transaction should happens in CDAP system context, not program context
-        ClassLoader oldClassLoader = ClassLoaders.setContextClassLoader(getClass().getClassLoader());
-        try {
-          txExecute(txContext, runnable);
-        } finally {
-          ClassLoaders.setContextClassLoader(oldClassLoader);
+        if (!responder.hasContentProducer()) {
+          contextReleaser.cancel();
         }
       }
-    };
+    }
   }
 
   /**
-   * Executes the given {@link TxRunnable} in a transaction. The transaction lifecycle is managed by the
-   * given {@link TransactionContext} instance.
-   *
-   * @throws TransactionFailureException if execution failed
-   * @throws TransactionConflictException if conflict is detected when trying to commit the transaction
+   * A {@link BodyProducerFactory} to be used when {@link #handleError(Throwable)} is called.
    */
-  private void txExecute(TransactionContext txContext, TxRunnable runnable) throws TransactionFailureException {
-    txContext.start();
-    try {
-      ClassLoader oldClassLoader = ClassLoaders.setContextClassLoader(programContextClassLoader);
-      try {
-        runnable.run(serviceContext);
-      } finally {
-        ClassLoaders.setContextClassLoader(oldClassLoader);
-      }
-    } catch (Throwable t) {
-      txContext.abort(new TransactionFailureException("Exception raised from TxRunnable.run()", t));
+  private static final class ErrorBodyProducerFactory implements BodyProducerFactory {
+
+    @Override
+    public BodyProducer create(HttpContentProducer contentProducer, TransactionalHttpServiceContext serviceContext) {
+      // It doesn't what it returns as it'll never get used
```

Comment:
I can in the current implementation. However, I didn't because the interface contract doesn't allow returning `null`. If any change in the implementation of the `DelayedHttpServiceResponder` that have this method called, then it's not good to have `null` returned.